Add CMap.ExtS.symmetric_diff

LasseBlaauwbroek opened the pull request:
`symmetric_diff` efficiently calculates the difference between two maps when they share most of their internal structure.

Note that this cannot be implemented on top of the public interface of CMap, as it relies on the internal implementation details. To implement this, Tactician currently accesses the Coq internals through `Obj.magic`, and Coq in turn accesses the OCaml internals through `Obj.magic`, which seems somewhat risky. So I'm hoping that this code can find a home here. But eventually I feel that the whole `CMap` extension to OCaml's map should be upstreamed there...

ppedrot commented on clib/cMap.ml:
```diff
@@ -37,6 +37,10 @@ sig
   val of_list : (key * 'a) list -> 'a t
+  val symmetric_diff :
+    eq:('a -> 'a -> bool) ->
+    (key -> [ `Left of 'a | `Right of 'a | `Unequal of 'a * 'a ] -> 'b -> 'b) ->
```
Don't use polymorphic variants, they're annoying as hell. You could define a common inductive in CSig or Utils, probably.

ppedrot replied:
Or better: just follow the Map.merge convention, which takes two 'a options. That would be even more regular.

LasseBlaauwbroek replied:
The problem with the `Map.merge` convention is that the API leaves the possibility for both `'a option`'s to be `None`, which can never happen. Are you sure you want to have that?

ppedrot replied:
I don't like Map.merge because of this, but uniformity is better than one bad API and one good API.

Alizter replied:
Also polyvariants like this will have an extra layer of boxing compared to regular variants.

SkySkimmer commented:
I think it may be nicer to have a `fold2`, then your `symmetric_diff` is just `fold2` where the callback uses the equality function to ignore symmetric parts.
(not sure `fold2` is quite the right name as eg `List.fold_left2` requires the lists have same length but this one wouldn't require the maps to have the same keys)

LasseBlaauwbroek commented:
@SkySkimmer the point of `symmetric_diff` is that it is much more efficient than `fold2`, because it is able to detect equal structure in the maps that don't have to be folded over. I'm not against a `fold2`, but that would not obviate the need for `symmetric_diff`.

SkySkimmer commented:
Your code only uses the `eq` in 
~~~coq
          if eq vl vr then rem
          else f kl (`Unequal (vl, vr)) rem
~~~
how is that different from passing `(fun k v acc -> match v with BothThere (a, b) -> if eq a b then acc else f k Unequal (a, b) acc | _ -> f k v acc)`?

SkySkimmer commented:
Also the name `symmetric_diff` seems like something that should produce a `Map.t` not a fold function, ie 
~~~ocaml
let symmetric_diff eq = merge (fun _ left right -> match left, right with 
  | None, None -> assert false
  | Some l, None -> Some (Left l)
  | None, Some r -> Some (Right r)
  | Some l, Some r -> if eq l r then Some (Both (l, r)) else None)
~~~

LasseBlaauwbroek commented:
Your proposed function would give the same results, but would be asymptotically slower. This happens due to checking of structural sharing.

Consider a very large map `m`. Now we create `let m' = Map.add k v m`. Then, the call `Map.symmetric_diff eq f m m' acc` will complete in `O(log (Map.size m))` time because the function only has to traverse to where the pair `(k, v)` are (and not are) in the maps. All other subtrees are structurally equal and can be ignored.

SkySkimmer commented on clib/cMap.ml:
```diff
@@ -193,6 +201,52 @@ struct
+    | MNode {l; v; d; r; _ } -> l, More (v, d, r, acc)
+
+  let rec symmetric_cons ((lm, la) as l) ((rm, ra) as r) =
+    if lm = rm then la, ra
```
oh there's a sneaky polymorphic equality here

LasseBlaauwbroek replied:
Good catch, I spotted it as well, but was waiting for the verdict on other requests to push :-) Fixed.

LasseBlaauwbroek commented:
Indeed `symmetric_diff` could output a `Map.t`, and I'm not super opposed to that. But with a fold you can be a little more efficient if your final datastructure is not a map. So I chose that...

ppedrot commented:
> Your proposed function would give the same results, but would be asymptotically slower. This happens due to checking of structural sharing.

I fail to see why. This is literally an inlining of the eq function. As a matter of fact, I also support removing eq from the arguments and calling this function fold2, with almost exactly the same interface as Map.merge.

ppedrot commented:
Also, writing it this way gives a straightforward spec: fold2 is the same as fold composed with merge, but more efficient.

LasseBlaauwbroek commented:
Ah, okay, I did not realize `fold2` would not fold over elements that occur in both sets.
I'm not sure if such an API is very intuitive for me. In such a function, `f k (Some v1) (Some v2)` might be called for any two values `v1` and `v2`, without any requirement on `v1` and `v2` being equal. But, the implementation is allowed to omit any pairs that are for certain known not to be equal. Or should the implementation exclude any two values that are not physically equal?

ppedrot commented:
> Or should the implementation exclude any two values that are not physically equal?

No, it's up to the argument function to decide what to do with the values. The two functions are equiexpressible, since you can get fold2 from the diff one by setting eq to be constantly false. But the fold2 function is easier to specify.

LasseBlaauwbroek commented:
Okay fine. But the name `fold2` is still problematic for me. From its name, I would expect `fold2` to either fold over all keys that occur in both maps or fold over the union of all the keys in either map. But folding over only the keys that occur in one map or the other (but not both, unless some special condition occurs), does not seem like `fold2` to me.

SkySkimmer commented:
>I fail to see why.

I think I figured it out, the important part is not so much the `eq` but rather the physical equality shortcuts in `symmetric_cons`

LasseBlaauwbroek commented:
> I think I figured it out, the important part is not so much the eq but rather the physical equality shortcuts in symmetric_cons

Right.

SkySkimmer commented:
We could replace the `eq` with physical equality, then the spec is "like `fold2`, but doesn't fold when both sides are physically equal"

LasseBlaauwbroek commented:
> We could replace the eq with physical equality, then the spec is "like fold2, but doesn't fold when both sides are physically equal"

That is fine with me.

LasseBlaauwbroek commented:
Okay, so summarizing, these appear to be the possible design questions:
- The signature of `f` can be either
  + `key -> [ `Left of 'a | `Right of 'a | `Unequal of 'a * 'a ] -> 'b -> 'b`
  + `key -> [ Left of 'a | Right of 'a | Unequal of 'a * 'a ] -> 'b -> 'b` (or some variant with different constructor names)
  + `key -> 'a option -> 'a option -> 'b -> 'b` which is consistent with `Map.merge` but has the caveat that `f k None None` will never occur
- We can either have `eq` as we have now, or replace instantiate `eq` with physical equality. In that case, people can still use a custom equality in `f`.
- Regarding the output: We can have `symmetric_diff` be a fold, or have it output a map. Or, we can rename the current function to `symmetric_diff_fold` and make `symmetric_diff` a specialization that outputs a map.
- Regarding the name: We could rename it to `fold2`, but that seems extremely counter-intuitive to me.

Can we get a consensus on this? I don't have a very strong opinion I guess.

SkySkimmer commented:
>Regarding the name: We could rename it to fold2, but that seems extremely counter-intuitive to me.

Definitely not, the renaming suggestion was based on a misunderstanding of what it does.

I like using physical equality instead of passing eq because then we don't have a reflexivity hypothesis for sane behaviour.

LasseBlaauwbroek commented:
- Renamed the function from `symmetric_diff` to `symmetric_diff_fold` (so that at a later point, version can be introduced that outputs a map).
- Removed the `eq` argument, and instantiated it with physical equality.
- The fold function now takes two optionals like @ppedrot requested. This API doesn't seem super clean to me though. You can see in hMap.ml that an `assert false` is needed to get around the impossible `None, None` case.

LasseBlaauwbroek commented:
Any volunteers to review/assign?

SkySkimmer commented:
@coqbot run full ci
